### PR TITLE
Include svg layers in screenshots.

### PR DESCRIPTION
**Pull request by manthey**
The screenshot function now returns a Promise.  It also can take an optional wait parameter to wait until the map is idle or just until a render animation frame has occurred.  An event is triggered when the screenshot is ready.

If we need it in the future, we can issue events during the preparation phase.

Update tests for promise-based screenshots.

A few of the selenium-based tests have been moved to the headless firefox-based tests.

This address some of issue #666.  Widget (html) layers and the attribution are not yet included.

### CODE REVIEW

**Comment by aashish24** [comment]
@manthey is this ready for review I am assuming?

**Comment by manthey** [comment]
@aashish24 I had thought so.  However, yesterday I added most of the support for arbitrary layers -- widget layers (even SVG) won't screenshot properly as it currently stands as not all css gets applied properly.  I should have an update later today that resolves all of that.

**Comment by manthey** [comment]
@aashish24 Ready for review.

This now fully resolves issue #666.

As a quick test in any of the examples, type the following in the console:
```
var map = $('#map').data('data-geojs-map');  // get a reference to the map
map.screenshot({wait:'idle'}).then(function (res) {  // take a screenshot
   $('body').append($('<img>').attr('src',res));  // when it is ready, add it to the page
   $('#map').css('display', 'none');  // and hide the original map
});
```

As mentioned in the last commit, screenshots of non-canvas layers will _not_ work in IE.

In any browser, fonts will look a bit different (slightly fuzzier), because they are rendered through SVG `foreignObject`, and SVG deliberately doesn't use hinting within the `foreignObject`.  There may be a way around that, but I haven't figured it out.

**Comment by aashish24** [comment]
Roger that @manthey I am on it now. 

> As a quick test in any of the examples, type the following in the console:

Cool! Let me build geojs in my new laptop and I will give it a shot.

File: src/map.js [hunk 1]
```diff
@@ -1547,20 +1547,46 @@ var map = function (arg) {
...
-    $a.appendTo(m_this.node());
+    /* Only add the element if there is at least one attribution */
+    if ($('span', $a).length) {
+      $a.appendTo(m_this.node());
+    }
     return m_this;
   };
 
   /**
+   * Draw a layer image to a canvas contex.  The layer's opacity and transform
```
Comment (aashish24) [inline src/map.js]: contex --> context?

File: src/map.js [hunk 2]
```diff
@@ -1571,14 +1597,53 @@ var map = function (arg) {
...
+      if (opts.wait === 'idle') {
+        m_this.onIdle(waitForRAF);
+      } else {
+        waitForRAF();
+      }
+      return defer.then(function () {
+        return m_this.screenshot(layers, type, encoderOptions, optsWithoutWait);
+      });
+    }
+    defer = $.when();
```
Comment (aashish24) [inline src/map.js]: do we ever return this resolved promise (since after this line we re-assign defer)?

Comment (aashish24) [inline src/map.js]: nm, I see the code below.

File: src/util/init.js [hunk 3]
```diff
@@ -687,6 +687,149 @@
...
+            styleElem.text(css);
+            fetch.resolve();
+          });
+          deferList.push(fetch);
+        }
+        $('head', container).append(styleElem);
+      });
+
+      $.when.apply($, deferList).then(function () {
+        var svg = $('<svg xmlns="http://www.w3.org/2000/svg"><foreignObject width="100%" height="100%"></foreignObject></svg>');
```
Comment (aashish24) [inline src/util/init.js]: @manthey should we wrap these long lines?

File: src/util/init.js [hunk 4]
```diff
@@ -687,6 +687,153 @@
     },
 
     /**
+     * Escape any character in a string that has a code point >= 127.
+     *
+     * @param {string} text: the string to escape.
+     * @returns {string}: the escaped string.
+     */
+    escapeUnicodeHTML: function (text) {
```
Comment (aashish24) [inline src/util/init.js]: @manthey where were some specific unicode char you observed?

Comment (manthey) [inline src/util/init.js]: The `&copy;` (&copy;) symbol in the attribution.  Generally, the SVG foreignObject seems to handle this as `&#169;`, but not as unescaped unicode.

**Comment by aashish24** [comment]
@manthey I see the HTML text is also getting blurry. Is that expected?

File: src/util/init.js (hunk 4 above)
Comment (aashish24) [inline src/util/init.js]: okay thanks!

**Comment by manthey** [comment]
Yes -- It can probably be addressed, but I'd rather defer that to another PR.

Basically, SVG `foreignObject` renders the HTML, and then we render the SVG as an image.  SVG doesn't ever use font hinting (so you don't get the pixel-targeted font variations, nor the subpixel anti-aliasing).  I don't know if rendering it at a higher resolution and down-sampling would help, or if there is a way to get it to use font-hinting.

**Comment by aashish24** [comment]
> Yes -- It can probably be addressed, but I'd rather defer that to another PR.

Sounds good to me. I do not know if there is an issue for this. If not can you create one?

**Comment by manthey** [comment]
Regarding font blurriness: are you on a platform that has a non-unity device scale?  For instance, on my Windows laptop, Windows is set to use 125% enlargement by default.  When Chrome or Firefox renders an image, it scales it UP so that it is 125% of the original size, which naturally makes it look blurry.  The original GeoJS render uses vector fonts at the appropriate resolution, but the generated screenshot is NOT the apparent resolution of the screen, but rather the scaled version.

This is probably the main component of whatever blurriness you are seeing.  What remains beyond that may be inconsequential (load the screenshot in something like GIMP, and you'll see that it is not the same as the resolution of the screen).

This means that we may want an option to scale by the device-pixel-ratio (or not), which would probably default to doing so.

**Pull request merged**